drivers: ethernet: stm32h7 IT based ethernet TX

lochej opened the pull request:
Modify the ethernet driver to use TX complete interrupts.
Adds HAL ethernet TX complete callback and locking semaphore.

Due to changing behavior/content of the TX DMA descriptors
on STM32H7 series, based on the state of the IP,
it is more reliable to wait for the TX complete interrupt to check
for DMA end of transmission event. This avoids polling the
DMA_DESC_OWN bit in the descriptors.

This improves reliability and performance of the ethernet peripheral.

Tested on CoapServer sample, Dumb HTTP server, telnet sample.

This implementation avoids all of the HAL_ETH_Transmit{Frame} failed
errors that occured with the current implementation on my
Nucleo-H743ZI board. Avoids system crashes and HAL_Timeout
of the original implementation.

@Nukersson tested approves this approach that was exposed in the
original ethernet driver PR #26226 

Benchmarked using the following: 
**Setup:** Killer E2400 Gigabit Ethernet
**Board:** Nucleo_H743ZI clocked from HSI 64 MHz
```
sudo ping -i 0.01 192.0.2.1 -c 1000 
--- 192.0.2.1 ping statistics ---
1000 packets transmitted, 1000 received, 0% packet loss, time 11822ms
rtt min/avg/max/mdev = 0.650/0.955/2.050/0.154 ms
```

Signed-off-by: Jeremy LOCHE <lochejeremy@gmail.com>

lochej committed:
drivers: ethernet: stm32h7 IT based ethernet TX

Modify the ethernet driver to use TX complete interrupts.
Adds HAL ethernet TX complete callback and locking semaphore.

Due to changing behavior/content of the TX DMA descriptors
on STM32H7 series, based on the state of the IP,
it is more reliable to wait for the TX complete interrupt to check
for DMA end of transmission event. This avoids polling the
DMA_DESC_OWN bit in the descriptors.

Improves reliability and performance of the ethernet peripheral.

Tested on CoapServer sample, Dumb HTTP server, telnet sample.

Signed-off-by: Jeremy LOCHE <lochejeremy@gmail.com>

Nukersson commented:
Please add @erwango also. I well look at your PR tomorrow also. Thank you for providing it. I hope, my prepared repository helped you also.

Nukersson commented:
Hey Jeremy. Thank you very much for your new approach. I like it. :)

As I see from the code you changed memory buffer sizes and introduced IT approach. Therefore I would split your PR into two following commits:

1. Changing the size of `tx_buffer_def`
2. Introducing interrupt approach

Please also try just to introduce the changes in commits descriptions briefly and/or with bullet points. Try also to avoid descriptions of advantages and disadvantages. Cause it is always clear, that every commit gains us some advantages ;) 

I've putted also some review from my side. You can always contact me by eMail, if I can help you.

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -210,9 +209,11 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
-	ETH_BufferTypeDef tx_buffer_def[ETH_TXBUFNB];
+	/* Only one tx_buffer is sufficient to pass only 1 dma_buffer */
+	/* We can save some memory here by allocating only 1 buffer */
+	ETH_BufferTypeDef tx_buffer_def[1];
```
I would use `ETH_TXBUFNB` define here and change it's value in code above to appropriate one. Please put this change into separate memory management commit.

lochej replied:
I have split into 2 commits, but ETH_TXBUFNB should be changed as they are used in the dma_tx_desc_tab. The ETH_BufferTypeDef from the HAL is just a linked list from which we can pass successive arrays to be transmitted. Here we only pass 1 dma_buffer corresponding to the first dma_buffer filled up with the net_buffer. So these symbols are differents.

Nukersson replied:
Thanks for splitting into 2 commits. Then introduce new one define with meaningful name (e.g. `ETH_TXBUF_DEF_LEN`) and value (1U in your case).

lochej replied:
I also thought of it in the sense that we only fill up 1 DMA descriptor at a time. 1 descriptor is used to transmit 1 ethernet frame. At the beginning of the function we exit if the size of the packet is bigger that the eth frame size. So this function needs a lot of work in order to send multiple frames at once. I will add the new symbol btw.

lochej replied:
@Nukersson  Changed to new define ETH_TXBUF_DEF_NB set to 1U.

Nukersson replied:
> I also thought of it in the sense that we only fill up 1 DMA descriptor at a time. 1 descriptor is used to transmit 1 ethernet frame. At the beginning of the function we exit if the size of the packet is bigger that the eth frame size. So this function needs a lot of work in order to send multiple frames at once. I will add the new symbol btw.

Same problem with budders I had also. I agree with you. We need only one tx buffer but not 4 as proposed in original driver. I think wi will fix it in this PR or open later a new one. As I said - I would test your PR later today.

jukkar replied:
> At the beginning of the function we exit if the size of the packet is bigger that the eth frame size.

Note that upper layer in network stack will not send bigger packets that can fit into Ethernet frame.

Nukersson replied:
> > At the beginning of the function we exit if the size of the packet is bigger that the eth frame size.
> 
> Note that upper layer in network stack will not send bigger packets that can fit into Ethernet frame.

Is it KConfig define then? Can you tell the name of this define, if yes?

jukkar replied:
> Is it KConfig define then? Can you tell the name of this define, if yes?

I do not understand what you mean by this. What defines are you looking for?

Nukersson replied:
> > Is it KConfig define then? Can you tell the name of this define, if yes?
> 
> I do not understand what you mean by this. What defines are you looking for?

I am looking for upper layer stack variable/define, which needed to configures packet's size for Ethernet frame.

jukkar replied:
For that you can use `NET_ETH_MAX_FRAME_SIZE`

Nukersson replied:
> > > Is it KConfig define then? Can you tell the name of this define, if yes?
> > 
> > 
> > I do not understand what you mean by this. What defines are you looking for?
> 
> I am looking for upper layer stack variable/define, which needed to configures packet's size for Ethernet frame.

Thank you very much.

Nukersson replied:
Thank you very much!

lochej replied:
> Same problem with budders I had also. I agree with you. We need only one tx buffer but not 4 as proposed in original driver. I think wi will fix it in this PR or open later a new one. As I said - I would test your PR later today.

@Nukersson I suggest to add this improvement into a new PR as the IT approach is a minimal change. Changing the handling of the buffer needs more time and testing to get working properly (as it plays with the net stack itself). We'll also have to consider checking for the availability of every DMA TX Descriptors before initiating a TX call because HAL_TransmitIT could fail if one of the DMA descriptors is not available.

Nukersson replied:
Ok. Sounds solide.

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -210,9 +209,11 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
+	ETH_BufferTypeDef tx_buffer_def[1];
 
-	memset(tx_buffer_def, 0, ETH_TXBUFNB*sizeof(ETH_BufferTypeDef));
+	memset(tx_buffer_def, 0, 1*sizeof(ETH_BufferTypeDef));
```
Please use `ETH_TXBUFNB` with appropriate defined value. Please put this change into separate memory management commit.

lochej replied:
Changed to new define ETH_TXBUF_DEF_NB set to 1U.

Nukersson replied:
Please push "Resolve conversation" button.

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -221,17 +222,37 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
 
 	if (hal_ret != HAL_OK) {
-		LOG_ERR("HAL_ETH_Transmit{Frame}: failed!");
+		LOG_ERR("HAL_ETH_TransmitFrame: failed! ret=%d err=%d", hal_ret, heth->ErrorCode);
```
I would just write: `HAL_ETH_Transmit: failed!`. The return values and error codes are not useful for the user, since they are not documented anywhere. The developers can see them from debug interface.

lochej replied:
Done 👍

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -221,17 +222,37 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
+	hal_ret = HAL_ETH_Transmit_IT(heth, &tx_config);
+
+	if (hal_ret != HAL_OK) {
+		LOG_ERR("HAL_ETH_TransmitIT: failed! ret=%d err=%d", hal_ret, heth->ErrorCode);
```
Please use `"HAL_ETH_TransmitIT: failed!"`.

lochej replied:
Done 👍

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -221,17 +222,37 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
+	/* So it is not freed before the interrupt happens */
+	hal_ret = HAL_ETH_Transmit_IT(heth, &tx_config);
+
+	if (hal_ret != HAL_OK) {
```
Putting it at line 253 ( after `#endif /* CONFIG_SOC_SERIES_STM32H7X */`) would be better. You can use then this error check for both H7 and non H7 versions of the driver. It is also better I think to put this error after next `k_sem_take` since taking semaphore is more important than logging error here!

lochej replied:
If for some reason the start of TX transmission fails, then we mustn't block on the ```k_sem_take``` call since the semaphore will never be given due to the absence of TX complete interrupt. This would lock the function forever and you would never exit from it. Therefore I dont agree with this proposition 🙃

Nukersson replied:
Ok. You're right. Please mark this conversation as resolved then.

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -221,17 +222,37 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
 
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
+
```
Do you need the two additional empty lines?

lochej replied:
Check out the updated commits ! 😄

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -485,6 +506,19 @@ static void eth_isr(void *arg)
+	__ASSERT_NO_MSG(dev_data != NULL);
+
+	k_sem_give(&dev_data->tx_int_sem);
+#endif /* CONFIG_SOC_SERIES_STM32H7X*/
```
Please put space before `*/`

lochej replied:
Done 👍

lochej committed:
drivers: ethernet: stm32h7 reduce count of tx_buffer_def

Reduced the size of tx_buffer_def array to 1 to save
on function stack memory. Here only 1 buffer is
enough to call the transmit function.

Signed-off-by: Jeremy LOCHE <lochejeremy@gmail.com>

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -52,7 +52,7 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #define ETH_MEDIA_INTERFACE_RMII	HAL_ETH_RMII_MODE
 
-
+#define ETH_TXBUF_DEF_NB (1U)
```
Using TAB would be better here. You also don't need the brackets as the expression is simple one.

Nukersson commented on drivers/ethernet/eth_stm32_hal_priv.h:
```diff
@@ -38,6 +38,10 @@ struct eth_stm32_hal_dev_data {
 	/* clock device */
 	struct device *clock;
+#ifdef CONFIG_SOC_SERIES_STM32H7X
+	/* interrupt semaphores to wait for ETH transfer complete IT */
```
Please delete this comment. It is obvious, what this semaphore do. I would also put this semaphore right after `rx_int_sem` (see line 46). Then it is clear - they are doing similar job.

lochej replied:
Done 👍

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -485,6 +498,19 @@ static void eth_isr(void *arg)
 	HAL_ETH_IRQHandler(heth);
 }
 
+void HAL_ETH_TxCpltCallback(ETH_HandleTypeDef *heth_handle)
```
Are you sure, the F4 and F7 series driver have same function name in it's HAL_ETH lib? Anyway we'll se it, after build kite were run.

lochej replied:
F4 and F7 both have this TX complete interrupt, buildkite won't fail with this function declared. Otherwise, I can just enclose it with the defines.

KozhinovAlexander reviewed (changes requested):
Added some small change requests

Nukersson commented:
Can't test this PR currently. Some of the commits in the Zephyr's master broke my board or clock. Investigating it now.

lochej commented:
> 
> 
> Can't test this PR currently. Some of the commits in the Zephyr's master broke my board or clock. Investigating it now.

Could it be #26980 ?

Nukersson commented:
> > Can't test this PR currently. Some of the commits in the Zephyr's master broke my board or clock. Investigating it now.
> 
> Could it be #26980 ?

Looks like this. I will switch now the conversation to #26980 to avoid message flood here.

xhpohanka commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -221,18 +221,32 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
+		goto error;
+	}
+
+	/* Wait for end of TX buffer transmission */
```
Just a question - would it make a sense to allocate the TX buffer(s) outside the stack, use sequentionally all available descriptors and release the memory inside the interrupt, instead of waiting here?

Nukersson replied:
> Just a question - would it make a sense to allocate the TX buffer(s) outside the stack, use sequentionally all available descriptors and release the memory inside the interrupt, instead of waiting here?

Thanks for good suggestion. But, I think it may not necessarily bring big performance gain, since:

1. Memory allocation/deallocation especially in interrupt may be performance consuming overhead.
2. In embedded systems it may be very helpful knowing the memory usage of whole system in advance, especially for the buffers, which are written/read permanently.

What we can do: Allocate doubled buffers of each type needed for one tx/rx operation. Thus we can perform eth action on say buffers set A while filling buffer set B. Then just switch them by setting correct pointer. Unfortunately it will increase memory usage, but may increase performance only a little.

lochej commented on drivers/ethernet/eth_stm32_hal_priv.h:
```diff
@@ -40,7 +40,10 @@ struct eth_stm32_hal_dev_data {
 	struct device *clock;
 	struct k_mutex tx_mutex;
 	struct k_sem rx_int_sem;
-	K_KERNEL_STACK_MEMBER(rx_thread_stack,
+#ifdef CONFIG_SOC_SERIES_STM32H7X
+	struct k_sem tx_int_sem;
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+	K_THREAD_STACK_MEMBER(rx_thread_stack,
 		CONFIG_ETH_STM32_HAL_RX_THREAD_STACK_SIZE);
 	struct k_thread rx_thread;
```
A commit brought new kernel stack definitions. I have resolved the merge conflict during my last rebase. This was tested and working on my Nucleo_H743ZI with HSI 64MHz as clock. PR is still valid.

lochej reviewed (commented):
Latest rebase had merge conflict with new thread stack definitions. I have resolved them in the last rebase.

lochej commented:
@ABOSTM @gmarull Would you be interested in reviewing this PR as you did with the original ethernet support PR (#26226) ?

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -209,9 +209,10 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
 
 #if defined(CONFIG_SOC_SERIES_STM32H7X)
-	ETH_BufferTypeDef tx_buffer_def[ETH_TXBUFNB];
+	/* Only one tx_buffer is sufficient to pass only 1 dma_buffer */
```
I would put this comment above the line with `#define ETH_TXBUF_DEF_NB	1U`.

lochej replied:
You're right that has more sense. Will do the change asap

lochej replied:
Done !

KozhinovAlexander reviewed (changes requested):
I've added one change request. Otherwise runs well on M7 core of nucleo_h745zi_q board.

lochej commented:
> 
> 
> I've added one change request. Otherwise runs well on M7 core of nucleo_h745zi_q board.

Your request has been fulfiled 👍 Please approve this PR or feel free to comment on anything else.

lochej commented:
Please @ABOSTM @gmarull @FRASTM review this PR. It should be ready for merging with your agreement 😄

lochej commented:
> Please @ABOSTM @gmarull @FRASTM @erwango  review this PR. It should be ready for merging with your agreement 😄

jukkar reviewed (approved):
LGTM

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -221,18 +221,32 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
+	}
+
+	/* Wait for end of TX buffer transmission */
+	if (k_sem_take(&dev_data->tx_int_sem, K_FOREVER) != 0) {
```
> Hello - I'm thinking about the `K_FOREVER` here. While testing this commit with Civetweb, I'm definitely able to get into situation when I have a deadlock. I still do not know what is the reason, I have a feeling that civetweb do not play well with current TCP stack.
> I tried to change it to some long timeout (20s) and at least it prevents (or more precisely give a chance to recover from) the deadlocks...

Are you using civetweb http-server?

xhpohanka replied:
> Are you using civetweb http-server?
Yes, I am. I have a simple site that provides a web interface to a sensor. There is an ajax routine that refreshes the information each 2s by request to a page returning short json string. When I wait long enough I usually see a deadlock here. I have not find exact steps to reproduce it though...

Nukersson replied:
> > Are you using civetweb http-server?
> 
> Yes, I am. I have a simple site that provides a web interface to a sensor. There is an ajax routine that refreshes the information each 2s by request to a page returning short json string. When I wait long enough I usually see a deadlock here. I have not find exact steps to reproduce it though...

Does the master's branch version of the driver work?

xhpohanka replied:
> Does the master's branch version of the driver work?

Hmmm, I need to check. But what do you mean by *master's branch version of the driver*? I think I have merged `lochej/drivers/stm32h7_it_based_eth` into my master, but it was probably very close.

Nukersson replied:
> > Does the master's branch version of the driver work?
> 
> Hmmm, I need to check. But what do you mean by *master's branch version of the driver*? I think I have merged `lochej/drivers/stm32h7_it_based_eth` into my master, but it was probably very close.

I mean [this one](https://github.com/zephyrproject-rtos/zephyr.git). It has currently h7 eth driver support implemented as close as possible to F7 one.

lochej replied:
I might have a solution to avoid having K_FOREVER here and keep a timeout. 

If you don't get the a TX complete interrupt that unlocks the semaphore, that means there must be some error that happened during DMA transmission or eth interrupt. I'll investigate on this issue but I was thinking about adding a timeout and semaphore reset to handle DMA transmission errors.

I'll be able to test later in the week but I'll update the PR so @Nukersson and @xhpohanka can test the error handling.

Nukersson replied:
> I might have a solution to avoid having K_FOREVER here and keep a timeout. 
> 
> If you don't get the a TX complete interrupt that unlocks the semaphore, that means there must be some error that happened during DMA transmission or eth interrupt. I'll investigate on this issue but I was thinking about adding a timeout and semaphore reset to handle DMA transmission errors.
> 
> I'll be able to test later in the week but I'll update the PR so @Nukersson and @xhpohanka can test the error handling.

Good idea. There should be DMA/ETH hardware error handlers in stm32h7 driver implementation. This would allow us handle described errors and/or drop/handle corrupted packets. I can take a look at posaible solution next week.

lochej replied:
@Nukersson @xhpohanka I've rolled out modifications to the driver to add the timeout to the semaphore and error logging. You can give it a try !

xhpohanka replied:
Thank you for the update @lochej. I do not have time to analyze it more today but FYI I still got `<err> eth_stm32_hal: HAL_ETH_TransmitIT tx_int_sem take timeout` quite often. Neither ETH nor DMA error is signaled.

xhpohanka replied:
@lochej here is a log of  timeouts,  I will probably share some more later, but now I changed settings a bit and I cannot reproduce it so easy like last week
```
[00:00:05.049,000] <err> eth_stm32_hal: HAL_ETH_TransmitIT tx_int_sem take timeout
[00:00:05.049,000] <err> eth_stm32_hal: eth packet timeout
d8 d0 90 52 8b 2e cc bd  35 ff ff ff 08 00 45 00 |...R.... 5.....E.
00 28 00 00 00 00 40 06  f5 c6 c0 a8 01 fb c0 a8 |.(....@. ........
01 be 93 5b 07 5b 2c 6f  c4 9a e5 b2 a1 af 50 10 |...[.[,o ......P.
05 00 12 a8 00 00                                |......           
[00:00:05.247,000] <inf> mqtt: MQTT client connected!
[00:00:05.264,000] <err> eth_stm32_hal: HAL_ETH_TransmitIT tx_int_sem take timeout
[00:00:05.265,000] <err> eth_stm32_hal: eth packet timeout
d8 d0 90 52 8b 2e cc bd  35 ff ff ff 08 00 45 00 |...R.... 5.....E.
00 28 00 00 00 00 40 06  f5 c6 c0 a8 01 fb c0 a8 |.(....@. ........
01 be 93 5b 07 5b 2c 6f  c4 b4 e5 b2 a1 b3 50 10 |...[.[,o ......P.
05 00 12 8a 00 00                                |......           
[00:00:05.917,000] <err> eth_stm32_hal: HAL_ETH_TransmitIT tx_int_sem take timeout
[00:00:05.917,000] <err> eth_stm32_hal: eth packet timeout
d8 d0 90 52 8b 2e cc bd  35 ff ff ff 08 00 45 00 |...R.... 5.....E.
00 44 00 00 00 00 40 01  f5 af c0 a8 01 fb c0 a8 |.D....@. ........
01 be 03 03 82 21 00 00  00 00 45 00 00 28 ed 8a |.....!.. ..E..(..
40 00 40 06 c8 3b c0 a8  01 be c0 a8 01 fb 07 5b |@.@..;.. .......[
80 0a 1c ce be dc 24 d8  1d 81 50 11 fa 1c 8b 43 |......$. ..P....C
00 00                                            |..               
[00:00:06.767,000] <inf> mqtt: PUBACK packet id: 56264
[00:00:08.283,000] <inf> mqtt: PUBACK packet id: 51096
[00:00:08.475,000] <err> eth_stm32_hal: HAL_ETH_TransmitIT tx_int_sem take timeout
[00:00:08.476,000] <err> eth_stm32_hal: eth packet timeout
d8 d0 90 52 8b 2e cc bd  35 ff ff ff 08 00 45 00 |...R.... 5.....E.
00 44 00 00 00 00 40 01  f5 af c0 a8 01 fb c0 a8 |.D....@. ........
01 be 03 03 82 21 00 00  00 00 45 00 00 28 ed 8c |.....!.. ..E..(..
40 00 40 06 c8 39 c0 a8  01 be c0 a8 01 fb 07 5b |@.@..9.. .......[
80 0a 1c ce be dc 24 d8  1d 81 50 11 fa 1c 8b 43 |......$. ..P....C
00 00                                            |..   
```

erwango commented on drivers/ethernet/eth_stm32_hal_priv.h:
```diff
@@ -40,7 +40,10 @@ struct eth_stm32_hal_dev_data {
+#ifdef CONFIG_SOC_SERIES_STM32H7X
+	struct k_sem tx_int_sem;
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+	K_THREAD_STACK_MEMBER(rx_thread_stack,
```
Should be K_KERNEL_STACK_MEMBER

lochej replied:
You're right my bad, I'm updating the PR, fixing push available now 👍

erwango commented:
@lochej Thanks again for this PR.
@Nukersson, @lochej I'm looking for active reviewers on STM32 components. Given your recent contribution in this area, would you be interested be added as codeowners for stm32 ethernet ? I know you mainly focus on H7, but this is a good start ;-)

Nukersson commented:
> @lochej Thanks again for this PR.
> @Nukersson, @lochej I'm looking for active reviewers on STM32 components. Given your recent contribution in this area, would you be interested be added as codeowners for stm32 ethernet ? I know you mainly focus on H7, but this is a good start ;-)

Looks good to me. It would be pleasure for me. And I really like to work with you (guys from ST) and Jeremy also!

lochej commented:
> @lochej Thanks again for this PR.
> @Nukersson, @lochej I'm looking for active reviewers on STM32 components. Given your recent contribution in this area, would you be interested be added as codeowners for stm32 ethernet ? I know you mainly focus on H7, but this is a good start ;-)

Of course that would be with great pleasure ! I have projects running on ST boards and I could sure help :)
